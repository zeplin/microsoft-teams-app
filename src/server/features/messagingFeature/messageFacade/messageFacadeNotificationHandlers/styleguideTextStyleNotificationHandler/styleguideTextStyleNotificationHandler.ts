import { StyleguideTextStyleEvent, WebhookEvent } from "../../../../../adapters/zeplin/types";
import { NotificationHandler } from "../NotificationHandler";
import { SHORT_DELAY } from "../constants";
import { commonTeamsCard, MessageCard } from "../teamsCardTemplates";
import { ZEPLIN_WEB_APP_BASE_URL, ZEPLIN_MAC_APP_URL_SCHEME } from "../../../../../config";
import { URL } from "url";

import { getMacAppRedirectURL } from "../getMacAppRedirectURL";

class StyleguideTextStyleNotificationHandler extends NotificationHandler {
    delay = SHORT_DELAY;
    private getText(events: StyleguideTextStyleEvent[]): string {
        const [{
            payload: {
                action,
                context: {
                    styleguide: {
                        name: styleguideName
                    }
                },
                resource: {
                    data: {
                        name: pivotTextStyleName
                    }
                }
            }
        }] = events;
        const actionText = action === "created" ? "added" : "updated";
        return events.length === 1
            ? `**${pivotTextStyleName}** is ${actionText} in _${styleguideName}_! 🏃‍♂`
            : `**${events.length} text styles** are ${actionText} in _${styleguideName}_! 🏃‍♂`;
    }

    private getWebappURL(
        events: StyleguideTextStyleEvent[]
    ): string {
        const [{
            payload: {
                context: {
                    styleguide: {
                        id: styleguideId
                    }
                }
            }
        }] = events;
        const webappURL = new URL(ZEPLIN_WEB_APP_BASE_URL);
        webappURL.pathname = `styleguide/${styleguideId}/styleguide/textstyles`;
        events.forEach(event => webappURL.searchParams.append("tsid", event.payload.resource.id));
        return webappURL.toString();
    }

    private getMacAppURL(
        events: StyleguideTextStyleEvent[]
    ): string {
        const [{
            payload: {
                context: {
                    styleguide: {
                        id: styleguideId
                    }
                }
            }
        }] = events;
        return getMacAppRedirectURL(`${ZEPLIN_MAC_APP_URL_SCHEME}://textStyles?stid=${styleguideId}&tsids=${events.map(event => event.payload.resource.id).join(",")}`);
    }

    shouldHandleEvent(event: WebhookEvent): boolean {
        return event.payload.action !== "deleted";
    }

    getTeamsMessage(
        events: StyleguideTextStyleEvent[]
    ): MessageCard {
        return commonTeamsCard({
            text: this.getText(events),
            section: {
                text: "Make sure your stylesheets are up to date!"
            },
            links: [{
                title: "Open in App",
                url: this.getMacAppURL(events)
            }, {
                title: "Open in Web",
                url: this.getWebappURL(events)
            }]
        });
    }
}

export const styleguideTextStyleNotificationHandler = new StyleguideTextStyleNotificationHandler();
