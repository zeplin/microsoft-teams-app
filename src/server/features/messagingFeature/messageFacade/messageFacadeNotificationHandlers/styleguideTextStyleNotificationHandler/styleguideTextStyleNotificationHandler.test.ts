import { styleguideTextStyleNotificationHandler } from "./styleguideTextStyleNotificationHandler";
import {
    StyleguideTextStyleCreateEvent,
    StyleguideTextStyleUpdateEvent
} from "../../../../../adapters/zeplin/types";

type GetDummyEventParams = {
    action?: string;
    textStyleId?: string;
    textStyleName?: string;
}

function getDummyEvent({
    action = "created",
    textStyleId = "textStyleId",
    textStyleName = "textStyleName"
}: GetDummyEventParams = {}): StyleguideTextStyleCreateEvent | StyleguideTextStyleUpdateEvent {
    return {
        payload: {
            action,
            context: {
                styleguide: {
                    id: "styleguideId",
                    name: "styleguideName"
                }
            },
            resource: {
                id: textStyleId,
                data: {
                    id: textStyleId,
                    name: textStyleName
                }
            }
        }
    } as StyleguideTextStyleCreateEvent | StyleguideTextStyleUpdateEvent;
}

describe("styleguideTextStyleNotificationHandler", () => {
    describe("getTeamsMessage method", () => {
        describe("for created notification", () => {
            it("should match snapshot when there is only 1 event", () => {
                expect(
                    styleguideTextStyleNotificationHandler.getTeamsMessage([getDummyEvent()])
                ).toMatchSnapshot();
            });

            it("should match snapshot when there are more than 1 events", () => {
                expect(
                    styleguideTextStyleNotificationHandler.getTeamsMessage([
                        getDummyEvent(),
                        getDummyEvent({
                            textStyleId: "anotherTextStyleId",
                            textStyleName: "anotherTextStyleName"
                        })
                    ])
                ).toMatchSnapshot();
            });
        });

        describe("for updated notification", () => {
            it("should match snapshot when there is only 1 event", () => {
                expect(
                    styleguideTextStyleNotificationHandler.getTeamsMessage([getDummyEvent({ action: "updated" })])
                ).toMatchSnapshot();
            });

            it("should match snapshot when there are more than 1 events", () => {
                expect(
                    styleguideTextStyleNotificationHandler.getTeamsMessage([
                        getDummyEvent({
                            action: "updated"
                        }),
                        getDummyEvent({
                            action: "updated",
                            textStyleId: "anotherTextStyleId",
                            textStyleName: "anotherTextStyleName"
                        })
                    ])
                ).toMatchSnapshot();
            });
        });
    });
});
