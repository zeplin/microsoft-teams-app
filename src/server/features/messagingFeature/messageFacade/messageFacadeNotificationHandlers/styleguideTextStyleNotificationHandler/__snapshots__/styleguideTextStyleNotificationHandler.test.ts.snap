// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styleguideTextStyleNotificationHandler getTeamsMessage method for created notification should match snapshot when there are more than 1 events 1`] = `
{
  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
  "type": "AdaptiveCard",
  "version": "1.2",
  "body": [
    {
      "type": "TextBlock",
      "text": "**2 text styles** are added in _styleguideName_! üèÉ‚Äç‚ôÇ",
      "wrap": true,
      "spacing": "small"
    },
    {
      "type": "Container",
      "spacing": "extraLarge",
      "items": [
        {
          "type": "TextBlock",
          "text": "Make sure your stylesheets are up to date!",
          "spacing": "none"
        }
      ]
    },
    {
      "type": "ActionSet",
      "spacing": "extraLarge",
      "actions": [
        {
          "type": "Action.OpenUrl",
          "title": "Open in App",
          "url": "zpl://textStyles?stid=styleguideId&tsids=textStyleId,anotherTextStyleId"
        },
        {
          "type": "Action.OpenUrl",
          "title": "Open in Web",
          "url": "https://app.zeplin.io/styleguide/styleguideId/styleguide/textstyles?tsid=textStyleId&tsid=anotherTextStyleId"
        }
      ]
    }
  ]
}
`;

exports[`styleguideTextStyleNotificationHandler getTeamsMessage method for created notification should match snapshot when there is only 1 event 1`] = `
{
  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
  "type": "AdaptiveCard",
  "version": "1.2",
  "body": [
    {
      "type": "TextBlock",
      "text": "**textStyleName** is added in _styleguideName_! üèÉ‚Äç‚ôÇ",
      "wrap": true,
      "spacing": "small"
    },
    {
      "type": "Container",
      "spacing": "extraLarge",
      "items": [
        {
          "type": "TextBlock",
          "text": "Make sure your stylesheets are up to date!",
          "spacing": "none"
        }
      ]
    },
    {
      "type": "ActionSet",
      "spacing": "extraLarge",
      "actions": [
        {
          "type": "Action.OpenUrl",
          "title": "Open in App",
          "url": "zpl://textStyles?stid=styleguideId&tsids=textStyleId"
        },
        {
          "type": "Action.OpenUrl",
          "title": "Open in Web",
          "url": "https://app.zeplin.io/styleguide/styleguideId/styleguide/textstyles?tsid=textStyleId"
        }
      ]
    }
  ]
}
`;

exports[`styleguideTextStyleNotificationHandler getTeamsMessage method for updated notification should match snapshot when there are more than 1 events 1`] = `
{
  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
  "type": "AdaptiveCard",
  "version": "1.2",
  "body": [
    {
      "type": "TextBlock",
      "text": "**2 text styles** are updated in _styleguideName_! üèÉ‚Äç‚ôÇ",
      "wrap": true,
      "spacing": "small"
    },
    {
      "type": "Container",
      "spacing": "extraLarge",
      "items": [
        {
          "type": "TextBlock",
          "text": "Make sure your stylesheets are up to date!",
          "spacing": "none"
        }
      ]
    },
    {
      "type": "ActionSet",
      "spacing": "extraLarge",
      "actions": [
        {
          "type": "Action.OpenUrl",
          "title": "Open in App",
          "url": "zpl://textStyles?stid=styleguideId&tsids=textStyleId,anotherTextStyleId"
        },
        {
          "type": "Action.OpenUrl",
          "title": "Open in Web",
          "url": "https://app.zeplin.io/styleguide/styleguideId/styleguide/textstyles?tsid=textStyleId&tsid=anotherTextStyleId"
        }
      ]
    }
  ]
}
`;

exports[`styleguideTextStyleNotificationHandler getTeamsMessage method for updated notification should match snapshot when there is only 1 event 1`] = `
{
  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
  "type": "AdaptiveCard",
  "version": "1.2",
  "body": [
    {
      "type": "TextBlock",
      "text": "**textStyleName** is updated in _styleguideName_! üèÉ‚Äç‚ôÇ",
      "wrap": true,
      "spacing": "small"
    },
    {
      "type": "Container",
      "spacing": "extraLarge",
      "items": [
        {
          "type": "TextBlock",
          "text": "Make sure your stylesheets are up to date!",
          "spacing": "none"
        }
      ]
    },
    {
      "type": "ActionSet",
      "spacing": "extraLarge",
      "actions": [
        {
          "type": "Action.OpenUrl",
          "title": "Open in App",
          "url": "zpl://textStyles?stid=styleguideId&tsids=textStyleId"
        },
        {
          "type": "Action.OpenUrl",
          "title": "Open in Web",
          "url": "https://app.zeplin.io/styleguide/styleguideId/styleguide/textstyles?tsid=textStyleId"
        }
      ]
    }
  ]
}
`;
